### PR TITLE
add sfdp pattern

### PR DESCRIPTION
**Pull request by yuchengliu1**
add sfdp pattern for MBartForCausalLM/PLBartForCausalLM in transformers==4.44.2. 
Improve the inference performance of these model.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 1]
```diff
@@ -581,6 +580,34 @@ def _sfdp_replacement_20(query, key, value, attn_mask, dropout_p):
         scale=1.0 / math.sqrt(query.size(-1)),
     )
 
+def _sfdp_pattern_21(query, key, value, attention_mask):
```
Comment (Valentine233) [inline torch/_inductor/fx_passes/fuse_attention.py]: Could you explain a bit about the difference between this pattern and the existing ones? Per my view, the reshapes ops are automatically generated.

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 2]
```diff
@@ -581,6 +580,34 @@ def _sfdp_replacement_20(query, key, value, attn_mask, dropout_p):
...
+    head_size= query.size(3)
+    q = query.view(bs * n_head, -1, head_size)
+    k = key.reshape(bs * n_head, -1, head_size)
+    v = value.reshape(bs * n_head, -1, head_size)
+    attn_weights = torch.bmm(q, k.transpose(1,2))
+    attn_weights = attn_weights.view(bs,n_head,seq_len,-1) + attention_mask
+    attn_weights = attn_weights.view(bs*n_head,seq_len,-1)
+    attn_weights = torch.nn.functional.softmax(attn_weights, dim=-1)
+    if query.dtype == torch.half:
+        attn_weights = attn_weights.to(torch.half)
```
Comment (Valentine233) [inline torch/_inductor/fx_passes/fuse_attention.py:602]: I suppose the convert dtype op can be automatically generated.

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 3]
```diff
@@ -581,6 +580,34 @@ def _sfdp_replacement_20(query, key, value, attn_mask, dropout_p):
...
+    attn_output = attn_output.view(bs, n_head, seq_len, head_size)
+    return attn_output
+
+def _sfdp_replacement_21(query, key, value, attention_mask):
+    counters["inductor"]["fuse_attention"] += 1
+    return _scaled_dot_product_attention(
+        query,
+        key,
+        value,
+        attn_mask=None,
```
Comment (Valentine233) [inline torch/_inductor/fx_passes/fuse_attention.py]: Input `attention_mask` is not used?

**Review by Valentine233 — COMMENTED** [review]
Please add a UT, and also the serialized pattern with https://github.com/pytorch/pytorch/blob/334a7d942ad2fb5e60c6208e1f56dcc65cd24018/torch/_inductor/fx_passes/serialized_patterns/_sfdp_pattern_18.py#L5-L6.

File: torch/_inductor/fx_passes/fuse_attention.py (hunk 1 above)
Comment (yuchengliu1) [reply torch/_inductor/fx_passes/fuse_attention.py]: It has 4d QKV transposed QKV input. however it use `bmm` in `Q @ K` and `attention @ V`. So it will reshape/view 4d QKV to 3d. It attention mask is also 4d, but its attention is 3d. So it will reshape attention to 4d first and add attention mask, then reshape back to 3d.

File: torch/_inductor/fx_passes/fuse_attention.py (hunk 2 above)
Comment (yuchengliu1) [reply torch/_inductor/fx_passes/fuse_attention.py:602]: This is a pattern for bf16/fp16. The convert will be automatically generated both before the add and after the softmax, or it will not be generated. However, I only want a convert after softmax.

File: torch/_inductor/fx_passes/fuse_attention.py (hunk 3 above)
Comment (yuchengliu1) [reply torch/_inductor/fx_passes/fuse_attention.py]: added attention_mask

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 4]
```diff
@@ -876,6 +904,13 @@ def _get_sfdp_patterns():
                 d,
                 _sfdp_extra_check(aten.div.Tensor),
             ),
+            (
+                _sfdp_pattern_21,
```
Comment (Valentine233) [inline torch/_inductor/fx_passes/fuse_attention.py]: `21` is not aligned with the generated/UT pattern number `24`.

File: torch/_inductor/fx_passes/fuse_attention.py (hunk 1 above)
Comment (Valentine233) [reply torch/_inductor/fx_passes/fuse_attention.py]: Please add some comments in codes.

**Review by Valentine233 — APPROVED** [review]

**Comment by yuchengliu1** [comment]
@drisspg. This PR is ready, please review this PR.

**Review by jansel — APPROVED** [review]
